src/Microsoft.AspNetCore.Server.Testing/Deployers/NginxDeployer.cs

```diff
@@ -0,0 +1,165 @@
+                ApplicationBaseUri = uri.ToString(),
+                HostShutdownToken = exitToken
+            };
+        }
+
+        private void SetupNginx(string redirectUri, Uri originalUri)
+        {
+            var startInfo = new ProcessStartInfo
+            {
+                FileName = "sudo",
+                Arguments = "-n true",
+                UseShellExecute = false,
+                CreateNoWindow = true,
+                RedirectStandardError = true,
+                RedirectStandardOutput = true,
+                // Trying a work around for https://github.com/aspnet/Hosting/issues/140.
+                RedirectStandardInput = true
+            };
+
+            using (var testSudo = new Process() { StartInfo = startInfo })
+            {
+                bool errorRecieved = false;
+                testSudo.ErrorDataReceived += (sender, dataArgs) =>
+                {
+                    if (!string.IsNullOrEmpty(dataArgs.Data))
+                    {
+                        errorRecieved = true;
+                    }
+                };
+                testSudo.Start();
+                testSudo.BeginErrorReadLine();
+                if (testSudo.WaitForExit(2000) == false)
+                {
+                    // process didn't exit
+                    throw new Exception("Failed to test for sudo");
+                }
+
+                if (errorRecieved == true)
+                {
+                    throw new InvalidOperationException("User doesn't have sudo permissions, can't deploy using Nginx");
+                }
+            }
+
+            // stop Nginx so that we can restart it with a new config file
+            startInfo.Arguments = "nginx -s stop";
+            using (var stopNginx = new Process() { StartInfo = startInfo })
+            {
+                stopNginx.ErrorDataReceived += (sender, dataArgs) =>
+                {
+                    if (!string.IsNullOrEmpty(dataArgs.Data))
+                    {
+                        Logger.LogWarning("nginx: " + dataArgs.Data);
+                    }
+                };
+                stopNginx.OutputDataReceived += (sender, dataArgs) =>
+                {
+                    if (!string.IsNullOrEmpty(dataArgs.Data))
+                    {
+                        Logger.LogInformation("nginx: " + dataArgs.Data);
+                    }
+                };
+                stopNginx.Start();
+                stopNginx.BeginErrorReadLine();
+                stopNginx.BeginOutputReadLine();
+                if (stopNginx.WaitForExit(2000) == false)
+                {
+                    // process didn't exit
+                    throw new Exception("Failed to stop Nginx");
+                }
+            }
+
+            // copy nginx.conf template and replace pertinent information
+            var configTemplateContent = File.ReadAllText(Path.Combine(DeploymentParameters.ApplicationPath, "nginx.conf"));
```

Comment:
No, it doesn't belong in Hosting, it would be wherever we use this deployer such as ServerTests